internal/k8s/kind.go

```diff
@@ -26,22 +28,56 @@ import (
 // objects, so we have to use a type assertion to detect kinds that
 // we care about.
 func KindOf(obj interface{}) string {
-	switch obj := obj.(type) {
-	case *v1.Secret:
-		return "Secret"
-	case *v1.Service:
-		return "Service"
-	case *v1.Endpoints:
-		return "Endpoints"
-	case *v1beta1.Ingress:
-		return "Ingress"
-	case *projectcontour.HTTPProxy:
-		return "HTTPProxy"
-	case *projectcontour.TLSCertificateDelegation:
-		return "TLSCertificateDelegation"
-	case *unstructured.Unstructured:
-		return obj.GetKind()
-	default:
-		return ""
+	gvk, _, err := scheme.Scheme.ObjectKinds(obj.(runtime.Object))
+	if err != nil {
+		switch obj := obj.(type) {
+		case *v1.Secret:
+			return "Secret"
+		case *v1.Service:
+			return "Service"
+		case *v1.Endpoints:
+			return "Endpoints"
+		case *v1beta1.Ingress:
+			return "Ingress"
+		case *projectcontour.HTTPProxy:
+			return "HTTPProxy"
+		case *projectcontour.TLSCertificateDelegation:
+			return "TLSCertificateDelegation"
+		case *unstructured.Unstructured:
+			return obj.GetKind()
+		default:
+			return ""
+		}
 	}
+	for _, gv := range gvk {
+		return gv.GroupKind().Kind
+	}
+	return ""
+}
+
+// VersionOf returns the GroupVersion string for the given Kubernetes object.
+//
+func VersionOf(obj interface{}) string {
+	//If err not nil we have the GVK and we can use it
+	//Else we have an error and we're going to use switch-case method as failover
+	//Some of the objects ex: HTTPProxy will get NotRegisteredErrForType errors so we have to use Else block
```

Comment:
Nit, could you please format this as a single paragraph:

```suggestion
	// If err is not nil we have the GVK and we can use it. Otherwise
	// we have an error and we're going to use switch-case method
	// as failover.  Some of the objects, e.g. HTTPProxy, will get
	// NotRegisteredErrForType errors so we have to the type assertion.

```

Reply:
done, also added the commit template